File: trees/tree_traversals.cpp
#include <vector>
#include <iostream>
#include <queue>
#include <stack>

using namespace std;

// Tree node
struct Node {
    int data;
    Node* left;
    Node* right;
};

// constructor for new tree node
Node* newNode(int data) {
    Node* node = new Node;
    node->data = data;
    node->left = NULL;
    node->right = NULL;
    return node;
}

//             1
//            / \
//           /   \
//          /     \
//         /       \
//        /         \
//      10           3
//     /  \         / \
//    /    \       /   \
//   4      6     2     9
//    \    / \   / \   / \
//    30  0   5 21  8 7  11
//

Node* constructTree() {

    Node* root               = newNode(1);
    root->left               = newNode(10);
    root->right              = newNode(3);

    root->left->left         = newNode(4);
    root->left->right        = newNode(6);

    root->right->left        = newNode(2);
    root->right->right       = newNode(9);

    root->left->left->right  = newNode(30);
    root->left->right->left  = newNode(0);
    root->left->right->right = newNode(5);

    root->right->left->left  = newNode(21);
    root->right->left->right = newNode(8);
    root->right->right->left = newNode(7);
    root->right->right->right= newNode(11);
    return root;
}

/**
 *
 */
void inOrderRecursive(Node* root) {
    if(root == NULL) {
        return;
    }

    inOrderRecursive(root->left);
    cout << root->data << " ";
    inOrderRecursive(root->right);
}

string inOrderRecursiveString(Node* root) {
    if(root == NULL) {
        return "";
    }

    string left = "";
    if(root->left) {
        left = inOrderRecursiveString(root->left);
    }

    string ans = to_string(root->data);

    string right = "";
    if(root->right) {
        right = inOrderRecursiveString(root->right);
    }

    return (left.empty() ? "" : left + " ") + ans + (right.empty() ? "" : " " + right);
}

void inOrderIterative(Node* root) {
    if(root == NULL) {
        return;
    }

    stack<Node*> s;
    Node* temp = root;
    s.push(temp);

    while(!s.empty()) {
        if(temp->left) {
            s.push(temp->left);
            temp = temp->left;
        } else {
            if(!s.empty()) {
                temp = s.top();
                s.pop();
                cout << temp->data << " ";
            }

            if(temp->right) {
                s.push(temp->right);
                temp = temp->right;
            }
        }
    }
}

void preOrderRecursive(Node* root) {
    if(root == NULL) {
        return;
    }

    cout << root->data << " ";
    preOrderRecursive(root->left);
    preOrderRecursive(root->right);
}

void preOrderIterative(Node* root) {
    if(root == NULL) {
        return;
    }

    Node* temp = root;
    stack<Node*> s;
    s.push(temp);

    while(!s.empty()) {
        if(!s.empty()) {
            temp = s.top();
            s.pop();
            cout << temp->data << " ";

            if(temp->right) {
                s.push(temp->right);
            }
            if(temp->left) {
                s.push(temp->left);
            }
        }
    }
}

void postOrderRecursive(Node* root) {

    if(root == NULL) {
        return;
    }

    postOrderRecursive(root->left);
    postOrderRecursive(root->right);
    cout << root->data << " ";
}

// PostOrder iterative is the hardest traversal out of inOrder, and preOrder
// Node* prev: keep track of the previously traversed node
// Node* curr: current node that’s on top of the stack
//
// When prev is curr‘s parent:
//         we are traversing down the tree
//         In this case, we try to traverse to curr‘s left child if available (push left child to stack)
//         If it is not available, we look at curr‘s right child
//         If both left and right child do not exist (curr is a leaf node), print curr and pop it off stack
//
// When prev is curr's left child:
//         we are traversing up the tree from the left
//         look at curr‘s right child. If available, then traverse down right child (push right child to stack)
//         If not available, print curr and pop it off stack
//
// When prev is curr‘s right child:
//         we are traversing up the tree from the right
//         print curr and pop it off stack
void postOrderIterative(Node* root) {
    if(root == NULL) {
        return;
    }

    // maintain parent pointer of root
    Node* prev = NULL;
    stack<Node*> s;
    s.push(root);

    while(!s.empty()) {
        Node* curr = s.top();
        if((prev == NULL) || (prev->left == curr) || (prev->right == curr)) {
            if(curr->left) {
                s.push(curr->left);
            } else if(curr->right) {
                s.push(curr->right);
            } else {
                cout << curr->data << " ";
                s.pop();
            }
        } else if(curr->left == prev) {
            if(curr->right) {
                s.push(curr->right);
            } else {
                cout << curr->data << " ";
                s.pop();
            }
        } else if(curr->right == prev) {
            cout << curr->data << " ";
            s.pop();
        }

        // set previous visited node for next iteration
        prev = curr;
    }
}

int height(Node* node) {
    if ( node == NULL) {
        return 0;
    } else {
        // compute the height of each subtree
        int lheight = height(node->left);
        int rheight = height(node->right);

        // use the larger one
        if (lheight > rheight) {
            return(lheight+1);
        } else {
            return(rheight+1);
        }
    }
}

// Print nodes at a given level
void printGivenLevel(Node* root, int level) {
    if (root == NULL) {
        return;
    }

    if (level == 1) {
        cout << root->data << " ";
    } else if (level > 1) {
        printGivenLevel(root->left, level-1);
        printGivenLevel(root->right, level-1);
    }
}

//print tree to check
void levelOrderRecursive(Node* root) {
    int h = height(root);
    for (int i = 1; i <= h; i++) {
        printGivenLevel(root, i);
    }
}

void levelOrderIterative(Node* root) {
    if(root == NULL) {
        return;
    }

    queue<Node*> myqueue;
    myqueue.push(root);
    int count = 0;

    while(!myqueue.empty()) {
        count = myqueue.size();
        while(count > 0) {
            Node* temp = myqueue.front();
            myqueue.pop();
            cout << temp->data << " ";

            if(temp->left != NULL) {
                myqueue.push(temp->left);
            }

            if(temp->right != NULL) {
                myqueue.push(temp->right);
            }
            count--;
        }
        cout << "| ";
    }
}

int num_nodes(Node* root) {

    int count_nodes = 0;
    if(root == NULL) {
        return count_nodes;
    }

    num_nodes(root->left);
    count_nodes++;
    num_nodes(root->right);
    return count_nodes;
}

void reverseLevelOrderIterative(Node* root) {

    if(root == NULL) {
        return;
    }

    vector<Node*> vec;
    vec.push_back(root);
    queue<Node*> myqueue;
    myqueue.push(root);
    int count = 0;

    while(!myqueue.empty()) {
        count = myqueue.size();
        while(count > 0) {
            Node* temp = myqueue.front();
            myqueue.pop();

            if(temp->left != NULL) {
                myqueue.push(temp->left);
                vec.push_back(temp->left);
            }

            if(temp->right != NULL) {
                myqueue.push(temp->right);
                vec.push_back(temp->right);
            }
            count--;
        }
    }

    reverse(vec.begin(), vec.end());
    for(int i = 0; i < vec.size(); i++) {
        cout << vec[i]->data << " ";
    }
}

int main() {

    Node* root = constructTree();

    int num_of_nodes = num_nodes(root);
    cout << "NUM NODES: " << num_of_nodes << endl;

    // inOrder
    cout << "InOrder Recursive:";
    inOrderRecursive(root);
    cout << endl;
    cout << "inOrder Recurisve string: " << inOrderRecursiveString(root) << endl;
    cout << "InOrder Iterative:";
    inOrderIterative(root);
    cout << endl;

    // preOrder
    cout << "PreOrder Recursive:";
    preOrderRecursive(root);
    cout << endl;
    cout << "PreOrder Iterative:";
    preOrderIterative(root);
    cout << endl;

    // postOrder
    cout << "PostOrder Recursive:";
    postOrderRecursive(root);
    cout << endl;
    cout << "PostOrder Iterative:";
    postOrderIterative(root);
    cout << endl;

    // level order 
    cout << "LevelOrder Recursive:";
    levelOrderRecursive(root);
    cout << endl;
    cout << "LevelOrder Iterative:";
    levelOrderIterative(root);
    cout << endl;
    cout << "Reverse LevelOrder Iterative:";
    reverseLevelOrderIterative(root);
    cout << endl;

    return 0;
}

File: trees/trees_py/tree_traversals.py
"""
Different methods of traversing a tree - inorder, preorder, and postorder
"""

# -*- encoding: utf-8 *-*


class Node:
    def __init__(self, data):
        self.data = data
        self.left = None
        self.right = None


class Tree:
    def __init__(self):
        self.root = None

    #           1
    #          / \
    #         /   \
    #        /     \
    #       /       \
    #      /         \
    #    10           3
    #   /  \         / \
    #  /    \       /   \
    # 4      6     2     9
    #  \    / \   / \   / \
    #  30  0   5 21  8 7  11

    def construct_tree(self):
        self.root = Node(1)
        self.root.left = Node(10)
        self.root.left.left = Node(4)
        self.root.left.left.right = Node(30)
        self.root.left.right = Node(6)
        self.root.left.right.left = Node(0)
        self.root.left.right.right = Node(5)
        self.root.right = Node(3)
        self.root.right.left = Node(2)
        self.root.right.left.left = Node(21)
        self.root.right.left.right = Node(8)
        self.root.right.right = Node(9)
        self.root.right.right.left = Node(7)
        self.root.right.right.right = Node(11)

    def height(self, root):
        if root is None:
            return 0
        left_height = self.height(root.left)
        right_height = self.height(root.right)

        if left_height > right_height:
            return left_height + 1
        else:
            return right_height + 1

    def in_order_traversal(self, root):
        if root is None:
            return
        self.in_order_traversal(root.left)
        print(root.data, end=" ")
        self.in_order_traversal(root.right)
        return

    def pre_order_traversal(self, root):
        if root is None:
            return
        print(root.data, end=" ")
        self.pre_order_traversal(root.left)
        self.pre_order_traversal(root.right)

    def post_order_traversal(self, root):
        if root is None:
            return
        self.post_order_traversal(root.left)
        self.post_order_traversal(root.right)
        print(root.data, end=" ")

    def print_given_level(self, root, level):
        """
            Given root of tree, print a level of the tree
        """
        if root is None:
            return

        if level == 1:
            print(root.data, end=" ")
        elif level > 1:
            self.print_given_level(root.left, level - 1)
            self.print_given_level(root.right, level - 1)

    def print_level_order_iterative(self, root):
        if root is None:
            return
        height = self.height(root)
        print()
        for i in range(1, height+1):
            self.print_given_level(root, i)
            print()

    def level_order_traversal(self, root):
        """
        Breadth-first traversal, print out the data by level
        """
        if root is None:
            return

        queue = []
        queue.append(root)
        print()
        while(len(queue) > 0):
            count = len(queue)
            while(count > 0):
                count -= 1
                curr_node = queue.pop(0)
                print(curr_node.data, end=" ")
                if curr_node.left:
                    queue.append(curr_node.left)
                if curr_node.right:
                    queue.append(curr_node.right)
            print()

    def level_order_traversal_reverse(self, root):
        """
        Breadth-first traversal, print out the data by level
        """
        if root is None:
            return

        queue = []
        queue.append(root)
        curr_pointer = 0
        total_nodes = self.get_node_count(root) - 1
        while(total_nodes > 0):
            count = len(queue)
            while(count > 0):
                count -= 1
                total_nodes -= 1
                curr_node = queue[curr_pointer]
                if curr_node.left:
                    queue.append(curr_node.left)
                    curr_pointer += 1
                if curr_node.right:
                    queue.append(curr_node.right)
                    curr_pointer += 1
            line_break_node = Node("new line")
            queue.append(line_break_node)

        print()
        for i in range(len(queue)):
            print(queue[i].data, end=" ")
        print()

    def get_node_count(self, root):
        if root is None:
            return 0

        return 1 + self.get_node_count(root.left) + self.get_node_count(root.right)


def main():
    tree = Tree()
    tree.construct_tree()

    print("in order traversal: ", end=" ")
    tree.in_order_traversal(tree.root)
    print()

    print("preorder traversal: ", end=" ")
    tree.pre_order_traversal(tree.root)
    print()

    print("postorder traversal: ", end=" ")
    tree.post_order_traversal(tree.root)
    print()

    print("level order traversal iterative", end=" ")
    tree.print_level_order_iterative(tree.root)
    print()

    print("total nodes in tree - ", tree.get_node_count(tree.root))
    print()

    print("level order traversal", end=" ")
    tree.level_order_traversal(tree.root)
    print()

    print("reverse level order traversal", end=" ")
    tree.level_order_traversal_reverse(tree.root)
    print()
    return

if __name__ == '__main__':
    main()
